src/tests/rules.rs

```diff
@@ -0,0 +1,63 @@
+#![cfg(test)]
+
+use std::sync::Arc;
+
+use proc_macros::{rule, rule_tests};
+
+use crate::{rule::Rule, violation, RuleTester};
+
+#[test]
+fn test_per_file_run_state() {
```

Comment:
This isn't robustly testing per-file-run state in and of itself (eg we should run against multiple files with `foo` in them and make sure it is keeping a separate "count" for each one, we should have a different fixing rule apply a fix and make sure this rule's count gets reset each time through the fixing loop)

But the scope of this PR ended up being more the per-file `tree_sitter_grep` callback-related stuff